pkg/compact/downsample/downsample.go

```diff
@@ -0,0 +1,585 @@
+
+	for _, s := range res {
+		s.close()
+	}
+	return res
+}
+
+func downsampleAggr(lset labels.Labels, all []sample, chunked [][]sample, window int64) (*series, error) {
+	if len(all) == 0 {
+		return nil, nil
+	}
+	metric := lset.Get("__name__")
+	ser := newSeries(lset, targetChunkSize(len(all))+1)
+
+	if strings.HasSuffix(metric, "$sum") {
+		downsampleSum(ser, all, window)
+	} else if strings.HasSuffix(metric, "$count") {
+		// Downsampling a count aggregate is equivalent to downsampling a sum.
+		downsampleSum(ser, all, window)
+	} else if strings.HasSuffix(metric, "$min") {
+		downsampleMin(ser, all, window)
+	} else if strings.HasSuffix(metric, "$max") {
+		downsampleMax(ser, all, window)
+	} else if strings.HasSuffix(metric, "$counter") {
+		downsampleCounterCounter(ser, all, chunked, window)
+	} else {
+		return nil, errors.Errorf("unknown aggregate metric %q", metric)
+	}
+	ser.close()
+	return ser, nil
+}
+
+func aggrLset(lset labels.Labels, aggr string) labels.Labels {
+	res := make(labels.Labels, len(lset))
+	copy(res, lset)
+
+	for i, l := range res {
+		if l.Name == "__name__" {
+			res[i].Value = fmt.Sprintf("%s$%s", l.Value, aggr)
+		}
+	}
+	return res
+}
+
+func isCounter(lset labels.Labels) bool {
+	metric := lset.Get("__name__")
+	return strings.HasSuffix(metric, "_total") ||
+		strings.HasSuffix(metric, "_bucket") ||
+		strings.HasSuffix(metric, "_sum")
+}
+
+func targetChunkSize(l int) int {
+	c := 1
+	for ; l/c > 250; c++ {
+	}
+	return l / c
+}
+
+type sample struct {
+	t int64
+	v float64
+}
+
+// series is a mutable series that creates XOR chunk with a configurable target amount
+// of samples per chunk.
+type series struct {
+	lset   labels.Labels
+	chunks []chunks.Meta
+
+	l, n       int
+	cur        chunkenc.Chunk
+	app        chunkenc.Appender
+	cmin, cmax int64
+}
+
+// newSeries returns a new series and automatically cuts a new chunk after n samples if
+// n is greater than 0.
+func newSeries(lset labels.Labels, n int) *series {
+	return &series{lset: lset, n: n}
+}
+
+// newSeries returns a new series and automatically cuts a new chunk after n samples if
+// n is greater than 0.
+// It extends the metric name of the label set by the aggregation modifier.
+func newAggrSeries(lset labels.Labels, aggr string, n int) *series {
+	return &series{lset: aggrLset(lset, aggr), n: n}
+}
+
+func (s *series) add(t int64, v float64) {
+	if s.n > 0 && s.l >= s.n {
+		s.cut()
+	}
+	if s.cur == nil {
+		s.cur = chunkenc.NewXORChunk()
+		s.cmin = t
+		s.app, _ = s.cur.Appender()
+	}
+	s.app.Append(t, v)
+	s.cmax = t
+}
+
+func (s *series) cut() {
+	s.chunks = append(s.chunks, chunks.Meta{
+		MinTime: s.cmin,
+		MaxTime: s.cmax,
+		Chunk:   s.cur,
+	})
+	s.cur = nil
+}
+
+func (s *series) close() {
+	if s.cur == nil {
+		return
+	}
+	s.chunks = append(s.chunks, chunks.Meta{
+		MinTime: s.cmin,
+		MaxTime: s.cmax,
+		Chunk:   s.cur,
+	})
+}
+
+// downsampleCounter downsamples the data points into the series with the given window.
```

Comment:
Mhm this comment (first sentence) could be more specific about `Counter` series here